haystack_experimental/dataclasses/chat_message.py

```diff
@@ -0,0 +1,187 @@
+    """
+
+    _role: ChatRole
+    _content: Sequence[ChatMessageContentT]
+    _name: Optional[str] = None
+    _meta: Dict[str, Any] = field(default_factory=dict, hash=False)
+
+    @property
+    def images(self) -> List[ImageContent]:
+        """
+        Returns the list of all images contained in the message.
+        """
+        return [content for content in self._content if isinstance(content, ImageContent)]
+
+    @property
+    def image(self) -> Optional[ImageContent]:
+        """
+        Returns the first image contained in the message.
+        """
+        if images := self.images:
+            return images[0]
+        return None
+
+    @classmethod
+    def from_user(
+        cls,
+        text: Optional[str] = None,
+        meta: Optional[Dict[str, Any]] = None,
+        name: Optional[str] = None,
+        *,
+        content_parts: Optional[Sequence[Union[TextContent, str, ImageContent]]] = None,
+    ) -> "ChatMessage":
+        """
+        Create a message from the user.
+
+        :param text: The text content of the message. Specify this or content_parts.
+        :param meta: Additional metadata associated with the message.
+        :param name: An optional name for the participant. This field is only supported by OpenAI.
+        :param content_parts: A list of content parts to include in the message. Specify this or text.
+        :returns: A new ChatMessage instance.
+        """
+        if not text and not content_parts:
+            raise ValueError("Either text or content_parts must be provided.")
+        if text and content_parts:
+            raise ValueError("Only one of text or content_parts can be provided.")
+
+        content: Sequence[Union[TextContent, ImageContent]] = []
+
+        if text is not None:
+            content = [TextContent(text=text)]
+        elif content_parts is not None:
+            content = [TextContent(el) if isinstance(el, str) else el for el in content_parts]
+
+        return cls(_role=ChatRole.USER, _content=content, _meta=meta or {}, _name=name)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Converts ChatMessage into a dictionary.
+
+        :returns:
+            Serialized version of the object.
+        """
+
+        serialized: Dict[str, Any] = {}
+        serialized["role"] = self._role.value
+        serialized["meta"] = self._meta
+        serialized["name"] = self._name
+        content: List[Dict[str, Any]] = []
+        for part in self._content:
+            if isinstance(part, TextContent):
+                content.append({"text": part.text})
+            elif isinstance(part, ToolCall):
+                content.append({"tool_call": asdict(part)})
+            elif isinstance(part, ToolCallResult):
+                content.append({"tool_call_result": asdict(part)})
+            elif isinstance(part, ImageContent):
+                content.append({"image": asdict(part)})
```

Comment:
I wonder if it would also make sense to have a way to serialize `ImageContent` by pointing to a `file_path` instead of it needing to be a base64_string. 

Similar to how you added a `from_file_path` utility method to `ImageContent` in your PoC I wonder if we should allow users to serialize and deserialize with this pointer instead. 

I'm thinking this b/c storing ChatMessages could become quite large in in disk if we can only store the base64 string and not optionally set a pointer instead. What do you think?

Reply:
Currently, I have the impression that serializing (and storing `ChatMessage`) is mostly used for `ChatPromptBuilder` and for multimodal inputs I would like to avoid so (as illustrated in the POC).

Did I miss something? Can you think of other use cases?

Reply:
> Can you think of other use cases?

I think some up and coming ones are short-term and long-term memory for Agent. Where we will probably revisit the ChatMessageWriter and ChatMessageRetriever in haystack-experimental. 

Additionally dC would have to decide how to handle this since we are basically writing the contents of ChatMessage to a dB to retain chat history. 

I think maybe a good alternative/workaround could be to handle in the `ChatMessageWriter` and `Retriever` how we'd like to handle ImageContent within ChatMessages. I.e. we could control there whether to skip writing the base64 string to the DB, but then I'd think we should store the file_path to the original image in the meta-field of ImageContent. What do you think?

Reply:
Thanks for the clarification, makes sense.

Handling this at the `ChatMessageWriter` and `Retriever` level sounds like a practical approach, and it helps keep the scope of the current multimodal work focused for now.

Storing a `file_path` in the `meta` field of `ImageContent` is feasible and would help keep the `ChatMessage` serialization logic simpler at this stage.

As we move forward, we can revisit/adjust this strategy.

WDYT?

Reply:
Sounds good to me!